src/Shared/HttpSys/NativeInterop/SocketAddress.cs

```diff
@@ -1,96 +1,73 @@
         {
-            return GetIpv6Address();
+            _sockaddr = sockaddr;
         }
-        else if (Family == AddressFamily.InterNetwork)
+
+        internal override int GetPort()
         {
-            return GetIPv4Address();
+            // sin_port is network byte order
+            return IPAddress.NetworkToHostOrder((short)_sockaddr.sin_port);
         }
-        else
+
+        internal override IPAddress? GetIPAddress()
         {
-            return null;
+            // address is network byte order
+            return new IPAddress(_sockaddr.sin_addr.S_un.S_addr);
         }
     }
 
-    private IPAddress GetIpv6Address()
-    {
-        Contract.Assert(_size >= IPv6AddressSize);
-        var bytes = new byte[NumberOfIPv6Labels * 2];
-        Array.Copy(_buffer, 8, bytes, 0, NumberOfIPv6Labels * 2);
-        return new IPAddress(bytes); // TODO: Does scope id matter?
-    }
-
-    private IPAddress GetIPv4Address()
+    private sealed class SocketAddressIPv6 : SocketAddress
     {
-        Contract.Assert(_size >= IPv4AddressSize);
-        return new IPAddress(new byte[] { _buffer[4], _buffer[5], _buffer[6], _buffer[7] });
-    }
+        private readonly SOCKADDR_IN6 _sockaddr;
 
-    internal static unsafe SocketAddress? CopyOutAddress(IntPtr address)
-    {
-        var addressFamily = *((ushort*)address);
-        if (addressFamily == (ushort)AddressFamily.InterNetwork)
+        internal SocketAddressIPv6(in SOCKADDR_IN6 sockaddr)
         {
-            var v4address = new SocketAddress(AddressFamily.InterNetwork, IPv4AddressSize);
-            fixed (byte* pBuffer = v4address._buffer)
-            {
-                for (var index = 2; index < IPv4AddressSize; index++)
-                {
-                    pBuffer[index] = ((byte*)address)[index];
-                }
-            }
-            return v4address;
+            _sockaddr = sockaddr;
         }
-        if (addressFamily == (ushort)AddressFamily.InterNetworkV6)
+
+        internal override int GetPort()
         {
-            var v6address = new SocketAddress(AddressFamily.InterNetworkV6, IPv6AddressSize);
-            fixed (byte* pBuffer = v6address._buffer)
-            {
-                for (var index = 2; index < IPv6AddressSize; index++)
-                {
-                    pBuffer[index] = ((byte*)address)[index];
-                }
-            }
-            return v6address;
+            // sin6_port is network byte order
+            return IPAddress.NetworkToHostOrder((short)_sockaddr.sin6_port);
         }
 
-        return null;
+        internal override IPAddress? GetIPAddress()
+        {
+            // address is network byte order
+            // when CsWin32 gets support for inline arrays, remove 'AsReadOnlySpan' call below.
+            // https://github.com/microsoft/CsWin32/issues/1086
+            return new IPAddress(_sockaddr.sin6_addr.u.Byte.AsReadOnlySpan()); // TODO: Does scope id matter?
```

Comment:
IPAddress does accept a scopeId, any reason no to just pass it in?

Reply:
TBH I don't know. The original code didn't pass it in, and I tried to avoid changing the observable behavior in this PR. (The "TODO" comment is copied from the original code.)